rust/src/builders/witness_builder.rs

```diff
@@ -277,86 +286,159 @@ impl TransactionWitnessSetBuilder {
     }
 
     pub fn add_existing(&mut self, wit_set: &TransactionWitnessSet) {
-        match &wit_set.vkeys() {
-            None => (),
-            Some(vkeys) => vkeys.0.iter().for_each(|vkey| { self.add_vkey(vkey); } ),
-        };
-        match &wit_set.bootstraps() {
-            None => (),
-            Some(bootstraps) => bootstraps.0.iter().for_each(|bootstrap| { self.add_bootstrap(bootstrap); } ),
-        };
-        match &wit_set.native_scripts() {
-            None => (),
-            Some(native_scripts) => native_scripts.0.iter().for_each(|native_script| { self.add_native_script(native_script); } ),
-        };
-        match &wit_set.plutus_scripts() {
-            None => (),
-            Some(plutus_scripts) => plutus_scripts.0.iter().for_each(|plutus_script| { self.add_plutus_script(plutus_script); } ),
-        };
-        match &wit_set.plutus_data() {
-            None => (),
-            Some(plutus_data) => plutus_data.elems.iter().for_each(|plutus_datum| { self.add_plutus_datum(plutus_datum); } ),
-        };
-        match &wit_set.redeemers() {
-            None => (),
-            Some(redeemers) => redeemers.0.iter().for_each(|redeemer| { self.add_redeemer(redeemer); } ),
-        };
+        if let Some(vkeys) = wit_set.vkeys() {
```

Comment:
Yeah it'd make more sense performance-wise to use `vkeys` directly and avoid the clone since we're using it from within the same crate. The clone on `vkeys()` is due to wasm since wasm_bindgen gets really ugly/confusing for users if you don't clone or if you take inputs by-value instead of by-ref since there is no rustc on the JS side telling you when something has been moved out or not, leading to confusing crashes if you're not careful or not aware of wasm_bindgen details (and I imagine most js users are not).

We end up cloning it later on in the function anyway with `add_vkey` so you're right there. In the current code it would be cloned twice.

In general performance isn't the hugest impact here since it's utterly and completely dwarfed by any cryptographic operations. If you profile it you'll see just a couple hash/sign functions taking up 95%+ of the entire process of using the tx builder. Avoiding recomputing hashes/etc is a much, much more important optimization for us.

I'm working on splitting up the library into a native rust version with separate wasm bindings so that's another reason why I wouldn't worry about performance issues that are using the wasm API in the tx builder right now. At first we might just only expose this in WASM but later on we'll migrate this code to be purely in rust just with a separate thin wasm_bindgen wrapper over top of it, so issues like this will get taken care of when we do that work.

Reply:
Extracting a pure rust crate is a good idea. Things like coin selection can be shared with Ergo or other eUTXO chain I presume? Do you need any help on that? @rooooooooob

Reply:
@siegfried we might need help later on after the main split. The phase I'm working on right now is updating the [`cddl-codegen`](https://github.com/Emurgo/cddl-codegen) tool which I initially used to generate most of the content inside of `serialization.rs`/`lib.rs`/`plutus.rs` in the first place. After re-generating the core structs via the new [`babbage.cddl`](https://github.com/input-output-hk/cardano-ledger/blob/master/eras/babbage/test-suite/cddl-files/babbage.cddl) binary spec for the new hardfork we should also support preserving abitrary CBOR encodings (since e.g. a CBOR length/int 5 could be inlined into the type or in 1/2/4/8 additional bytes, there are 2 ways to encode lengths (definite/indefinite followed by a break code), maps can vary in order, etc) as well as the split into a rust crate and a wasm wrapper crate instead of trying to support both at once. This should make the rust user experience a lot less awful.

After I do this main split via `cddl-codegen` we'll still need to adapt a lot of the handwritten code (in particular `tx_builder.rs`, `utils.rs`, etc but also some of the hand-edits to `lib.rs`) afterwards which is where there might be some work to be done, especially in updating the tx builder to be just in rust. We'll have to see how this comes out. I hope to start the regeneration process sometime next week. In most cases it's just passing the spec through the codegen but there are some things that either a) aren't well supported or b) generate awkward APIs so those I'll need to do by hand.

Reply:
> since wasm_bindgen gets really ugly/confusing for users if you don't clone or if you take inputs by-value instead of by-ref since there is no rustc on the JS side telling you when something has been moved out or not, leading to confusing crashes if you're not careful or not aware of wasm_bindgen details (and I imagine most js users are not).

Is it because the argument given to the function is a reference?

Reply:
I presume the functions exposed as WASM should use the object instead of reference?